#### Name: Andrew Shapiro
#### Date: 11/26/2024
#### Exercise: Milestone #2: Final Project
#### Description: This notebook will read in my Project 3 Dataset on Board Games and explore the dataset using some of pandas' properties and built-in functions.

In [1]:
import pandas as pd

## 1. Using a Jupyter Notebook, import the data into a pandas dataframe using the appropriate Python library for the dataset file type.
For this assignment, I have a huge interest in board games, so that's what my final project is about.
* My dataset is in a file called `BGG.csv`. The link to the original website for the dataset is https://ieee-dataport.org/open-access/boardgamegeek-dataset-board-games.

In [4]:
# Read the dataset into the pandas data frame.
df = pd.read_csv('BGG.csv')

## 2. Explore the dataset using some of pandas’ properties and built-in functions.
For this assignment, I will show my explorations and results in separate notebook cells. The list below are all of the pandas properties and built-in functions I will be exploring. 
* describe()
* columns
* shape
* dtypes
* head(), tail(), sample()
* info()

### **(a)** df.describe()
After I ran this function on the dataframe, I was given a summary of all questions in the dataframe. I was able to see the mean age for all board games, the average rating, and complexitity! I was also able to see all of the percentiles (25%, 50%, and 75%). I notice that only the columns with number values were shown, and not the string. I believe this function will only work on columns that have number values, and not strings, which makes a lot of sense considering that the values it outputs will only work on numbers.

In [5]:
df.describe()

ID  Year Published   Min Players   Max Players  \
count   20327.000000    20342.000000  20343.000000  20343.000000   
mean   108216.245142     1984.249877      2.019712      5.672221   
std     98682.097298      214.003181      0.690366     15.231376   
min         1.000000    -3500.000000      0.000000      0.000000   
25%     11029.000000     2001.000000      2.000000      4.000000   
50%     88931.000000     2011.000000      2.000000      4.000000   
75%    192939.500000     2016.000000      2.000000      6.000000   
max    331787.000000     2022.000000     10.000000    999.000000   

          Play Time       Min Age    Users Rated  Rating Average  \
count  20343.000000  20343.000000   20343.000000    20343.000000   
mean      91.294548      9.601485     840.971391        6.403227   
std      545.447203      3.645458    3511.562220        0.935911   
min        0.000000      0.000000      30.000000        1.050000   
25%       30.000000      8.000000      55.000000        5.820000   
50%       45.000000     10.000000     120.000000        6.430000   
75%       90.000000     12.000000     385.000000        7.030000   
max    60000.000000     25.000000  102214.000000        9.580000   

           BGG Rank  Complexity Average    Owned Users  
count  20343.000000        20343.000000   20320.000000  
mean   10172.890429            1.991188    1408.457628  
std     5872.831597            0.848903    5040.179315  
min        1.000000            0.000000       0.000000  
25%     5087.500000            1.330000     146.000000  
50%    10173.000000            1.970000     309.000000  
75%    15258.500000            2.540000     864.000000  
max    20344.000000            5.000000  155312.000000

### **(b)** df.columns
After I ran this property on the dataframe, I was given an object filled with strings with all of the column names in the dataset.

In [6]:
df.columns

Index(['ID', 'Name', 'Year Published', 'Min Players', 'Max Players',
       'Play Time', 'Min Age', 'Users Rated', 'Rating Average', 'BGG Rank',
       'Complexity Average', 'Owned Users', 'Mechanics', 'Domains'],
      dtype='object')

### **(c)** df.shape
After I ran this property on the dataframe, it returned a tuple that displayed the shape of the dataset in rows, columns format. I found that there were 20343 observations (types of board games) and there were 14 columns in this dataset using this property.

In [7]:
df.shape

(20343, 14)

### **(d)** df.dtypes
After I ran this property on the dataframe, it returned the data type of every column of the dataset. With this command, I noticed that a majority of the dataset was an 'float64' or 'int64' datatype, which means that a large part of this dataset is number based. I also noticed that all of my string related column information is in an 'object' datatype.

In [8]:
df.dtypes

ID                    float64
Name                   object
Year Published        float64
Min Players             int64
Max Players             int64
Play Time               int64
Min Age                 int64
Users Rated             int64
Rating Average        float64
BGG Rank                int64
Complexity Average    float64
Owned Users           float64
Mechanics              object
Domains                object
dtype: object

### **(e)** df.head()
After I ran this function on the dataframe, I was given the first five responses in the dataset. With this command, I was able to see a glimpse of the dataset using the first five rows as a sample.

In [9]:
df.head()

ID                               Name  Year Published  Min Players  \
0  174430.0                         Gloomhaven          2017.0            1   
1  161936.0          Pandemic Legacy: Season 1          2015.0            2   
2  224517.0                  Brass: Birmingham          2018.0            2   
3  167791.0                  Terraforming Mars          2016.0            1   
4  233078.0  Twilight Imperium: Fourth Edition          2017.0            3   

   Max Players  Play Time  Min Age  Users Rated  Rating Average  BGG Rank  \
0            4        120       14        42055            8.79         1   
1            4         60       13        41643            8.61         2   
2            4        120       14        19217            8.66         3   
3            5        120       12        64864            8.43         4   
4            6        480       14        13468            8.70         5   

   Complexity Average  Owned Users  \
0                3.86      68323.0   
1                2.84      65294.0   
2                3.91      28785.0   
3                3.24      87099.0   
4                4.22      16831.0   

                                           Mechanics  \
0  Action Queue, Action Retrieval, Campaign / Bat...   
1  Action Points, Cooperative Game, Hand Manageme...   
2  Hand Management, Income, Loans, Market, Networ...   
3  Card Drafting, Drafting, End Game Bonuses, Han...   
4  Action Drafting, Area Majority / Influence, Ar...   

                          Domains  
0  Strategy Games, Thematic Games  
1  Strategy Games, Thematic Games  
2                  Strategy Games  
3                  Strategy Games  
4  Strategy Games, Thematic Games

In [10]:
# I noticed that you can customize the number of samples so I tried it out here!
df.head(3)

ID                       Name  Year Published  Min Players  \
0  174430.0                 Gloomhaven          2017.0            1   
1  161936.0  Pandemic Legacy: Season 1          2015.0            2   
2  224517.0          Brass: Birmingham          2018.0            2   

   Max Players  Play Time  Min Age  Users Rated  Rating Average  BGG Rank  \
0            4        120       14        42055            8.79         1   
1            4         60       13        41643            8.61         2   
2            4        120       14        19217            8.66         3   

   Complexity Average  Owned Users  \
0                3.86      68323.0   
1                2.84      65294.0   
2                3.91      28785.0   

                                           Mechanics  \
0  Action Queue, Action Retrieval, Campaign / Bat...   
1  Action Points, Cooperative Game, Hand Manageme...   
2  Hand Management, Income, Loans, Market, Networ...   

                          Domains  
0  Strategy Games, Thematic Games  
1  Strategy Games, Thematic Games  
2                  Strategy Games

### **(f)** df.tail()
After I ran this function on the dataframe, I was given the last five responses in the dataset. With this command, I was able to see a glimpse of the dataset using the last five rows as a sample.

In [11]:
df.tail()

ID                Name  Year Published  Min Players  Max Players  \
20338  16398.0                 War             0.0            2            2   
20339   7316.0               Bingo          1530.0            2           99   
20340   5048.0          Candy Land          1949.0            2            4   
20341   5432.0  Chutes and Ladders          -200.0            2            6   
20342  11901.0         Tic-Tac-Toe         -1300.0            2            2   

       Play Time  Min Age  Users Rated  Rating Average  BGG Rank  \
20338         30        4         1340            2.28     20340   
20339         60        5         2154            2.85     20341   
20340         30        3         4006            3.18     20342   
20341         30        3         3783            2.86     20343   
20342          1        4         3275            2.68     20344   

       Complexity Average  Owned Users  \
20338                1.00        427.0   
20339                1.05       1533.0   
20340                1.08       5788.0   
20341                1.02       4400.0   
20342                1.16       1374.0   

                                               Mechanics  \
20338                                                NaN   
20339   Betting and Bluffing, Bingo, Pattern Recognition   
20340                               Roll / Spin and Move   
20341  Dice Rolling, Grid Movement, Race, Roll / Spin...   
20342                 Paper-and-Pencil, Pattern Building   

                                Domains  
20338                  Children's Games  
20339                       Party Games  
20340                  Children's Games  
20341                  Children's Games  
20342  Abstract Games, Children's Games

In [12]:
# I noticed that you can customize the number of samples so I tried it out here!
df.tail(3)

ID                Name  Year Published  Min Players  Max Players  \
20340   5048.0          Candy Land          1949.0            2            4   
20341   5432.0  Chutes and Ladders          -200.0            2            6   
20342  11901.0         Tic-Tac-Toe         -1300.0            2            2   

       Play Time  Min Age  Users Rated  Rating Average  BGG Rank  \
20340         30        3         4006            3.18     20342   
20341         30        3         3783            2.86     20343   
20342          1        4         3275            2.68     20344   

       Complexity Average  Owned Users  \
20340                1.08       5788.0   
20341                1.02       4400.0   
20342                1.16       1374.0   

                                               Mechanics  \
20340                               Roll / Spin and Move   
20341  Dice Rolling, Grid Movement, Race, Roll / Spin...   
20342                 Paper-and-Pencil, Pattern Building   

                                Domains  
20340                  Children's Games  
20341                  Children's Games  
20342  Abstract Games, Children's Games

### **(g)** df.sample()
After I ran this function on the dataframe, I was given a random observation inside of the dataset. However, I noticed that I was able to give it a parameter, so I was able to ask it for multiple random observations inside of the dataset as well. This was fun to play around with! I also applied the same parameter feature to the above commands as well.

In [13]:
df.sample()

ID                                 Name  Year Published  \
4569  220203.0  Purrrlock Holmes: Furriarty's Trail          2017.0   

      Min Players  Max Players  Play Time  Min Age  Users Rated  \
4569            2            5         30       10          430   

      Rating Average  BGG Rank  Complexity Average  Owned Users  \
4569            6.58      4571                 1.6       1058.0   

                        Mechanics       Domains  
4569  Memory, Pattern Recognition  Family Games

In [15]:
# I noticed that you can customize the number of samples so I tried it out here!
df.sample(5)

ID              Name  Year Published  Min Players  Max Players  \
12091   21506.0    New Mastermind          2004.0            2            5   
2242   216403.0           Element          2017.0            2            4   
16834   27674.0        Nacho Loco          2007.0            2            6   
19282   84825.0         FaceEater          2010.0            2            6   
18005   87748.0  Frenetic Village          2010.0            2            8   

       Play Time  Min Age  Users Rated  Rating Average  BGG Rank  \
12091         30        8          209            5.79     12093   
2242          60       14          793            7.29      2244   
16834         20        8          199            5.56     16836   
19282         60       15           69            4.60     19284   
18005         20        8           45            5.32     18007   

       Complexity Average  Owned Users  \
12091                1.88        623.0   
2242                 2.33       1415.0   
16834                1.00        457.0   
19282                2.75        215.0   
18005                1.33         47.0   

                                               Mechanics         Domains  
12091                                                NaN  Abstract Games  
2242   Area Majority / Influence, Enclosure, Grid Mov...  Abstract Games  
16834                                     Tile Placement             NaN  
19282                                     Set Collection             NaN  
18005                                      Card Drafting             NaN

### **(h)** df.info()
After I ran this function on the dataframe, I noticed that it gave me the list of columns again and also gave me the column's corresponding index. Additionally, I noticed that it also gave the 'class', RangeIndex, DataTypes, and memory usage. I noticed now, more clearly, that a lot of the columns in this dataset was of a number datatype (float64, int64) and the 2 string columns (which is supported by our .dtypes() function).

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20327 non-null  float64
 1   Name                20343 non-null  object 
 2   Year Published      20342 non-null  float64
 3   Min Players         20343 non-null  int64  
 4   Max Players         20343 non-null  int64  
 5   Play Time           20343 non-null  int64  
 6   Min Age             20343 non-null  int64  
 7   Users Rated         20343 non-null  int64  
 8   Rating Average      20343 non-null  float64
 9   BGG Rank            20343 non-null  int64  
 10  Complexity Average  20343 non-null  float64
 11  Owned Users         20320 non-null  float64
 12  Mechanics           18745 non-null  object 
 13  Domains             10184 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 2.2+ MB
